### PR TITLE
[KSECDD] Fix RtlEncryptMemory test regression

### PR DESCRIPTION
**Pull request by tkreuzer**
Fixes use of uninitialized memory.

### CODE REVIEW

**Review by binarymaster — APPROVED** [review]

**Review by HBelusca — DISMISSED** [review]
I won't approve this since it seems that you revert completely unrelated things to what the RtlEncryptMemory test (and the function itself in advapi32) uses ...

**Comment by tkreuzer** [comment]
> 
> 
> I won't approve this since it seems that you revert completely unrelated things to what the RtlEncryptMemory test (and the function itself in advapi32) uses ...

LOL. Ok, so let's keep it broken then...

**Comment by learn-more** [comment]
seems to be gcc only.
Since it likes to bug out at random boots, let's just revert this and be done with it.

**Comment by SergeGautherie** [comment]
Update msvcrt.spec too?

File: drivers/crypto/ksecdd/crypt.c [hunk 1]
```diff
@@ -27,6 +27,18 @@ typedef struct _KSEC_LOGON_DATA
...
+void PrintKeyData(PUCHAR KeyData)
+{
+    ULONG i;
+    for (i = 0; i < 32; i++)
+    {
+        DbgPrint("%02X", KeyData[i]);
+    }
+    DbgPrint("\n");
+}
+#endif
```
Comment (HBelusca) [inline drivers/crypto/ksecdd/crypt.c:40]: May be removed in the final commit? (or do you want to keep it for reference purposes?)

File: modules/rostests/apitests/advapi32/RtlEncryptMemory.c [hunk 2]
```diff
@@ -9,26 +9,118 @@
...
     ok_int(RtlEqualMemory(Buffer, TestData, sizeof(Buffer)), 0);
-    Status = RtlDecryptMemory(Buffer, sizeof(Buffer), RTL_ENCRYPT_OPTION_SAME_PROCESS);
+    Status = RtlDecryptMemory(Buffer, sizeof(Buffer), OptionFlags);
     ok_ntstatus(Status, STATUS_SUCCESS);
     ok_int(RtlEqualMemory(Buffer, TestData, sizeof(Buffer)), 1);
+}
+
+
+
+void TestCrossProcessDecrypt(PSTR Param)
```
Comment (HBelusca) [inline modules/rostests/apitests/advapi32/RtlEncryptMemory.c:46]:
```suggestion
}

void TestCrossProcessDecrypt(PSTR Param)
```

File: modules/rostests/apitests/advapi32/RtlEncryptMemory.c [hunk 3]
```diff
@@ -9,26 +9,118 @@
...
+    }
+
+    printf("TestCrossProcessDecrypt(%s)\n", StrBuffer - 5);
+
+    system(CmdLine);
+}
+
+START_TEST(RtlEncryptMemory)
+{
+    CHAR _CRT_ALIGN(8) Buffer[MAX_PATH] = { 0 };
```
Comment (HBelusca) [inline modules/rostests/apitests/advapi32/RtlEncryptMemory.c:104]: `DECLSPEC_ALIGN(8)` ?

File: drivers/crypto/ksecdd/crypt.c (hunk 1 above)
Comment (tkreuzer) [reply drivers/crypto/ksecdd/crypt.c:40]: I'd like to keep it for debugging.

File: modules/rostests/apitests/advapi32/RtlEncryptMemory.c [hunk 4]
```diff
@@ -9,26 +9,117 @@
...
+    RtlCopyMemory(Buffer, TestData, sizeof(Buffer));
+    Status = RtlEncryptMemory(Buffer, 7, OptionFlags);
     ok_ntstatus(Status, STATUS_INVALID_PARAMETER);
 
-    /* Buffer must not be aligned to 8 bytes */
-    Status = RtlEncryptMemory(&Buffer[1], 8, RTL_ENCRYPT_OPTION_SAME_PROCESS);
+    /* Buffer can be unaligned */
+    RtlCopyMemory(Buffer, TestData, sizeof(Buffer));
+    Status = RtlEncryptMemory(&Buffer[1], 8, OptionFlags);
+    ok_ntstatus(Status, STATUS_SUCCESS);
```
Comment (SergeGautherie) [inline modules/rostests/apitests/advapi32/RtlEncryptMemory.c:27]: Everywhere, add a check that `Buffer` content did change.

File: modules/rostests/apitests/advapi32/RtlEncryptMemory.c [hunk 5]
```diff
@@ -9,26 +9,117 @@
...
+static const CHAR TestData[32] = "This is some test Message!!!";
+
+void TestEncrypt(ULONG OptionFlags)
 {
-    static const CHAR TestData[32] = "This is some test Message!!!";
-    CHAR Buffer[32];
+    CHAR DECLSPEC_ALIGN(8) Buffer[32];
     NTSTATUS Status;
 
     /* Size must be aligned to 8 bytes */
```
Comment (SergeGautherie) [inline modules/rostests/apitests/advapi32/RtlEncryptMemory.c:19]:
```suggestion
    /* Size must be a multiple of RTL_ENCRYPT_MEMORY_SIZE */
```

File: modules/rostests/apitests/advapi32/RtlEncryptMemory.c [hunk 6]
```diff
@@ -9,26 +9,117 @@
...
+    printf("TestCrossProcessDecrypt(%s)\n", Param);
+
+    OptionFlags = Param[3] == '1' ? RTL_ENCRYPT_OPTION_CROSS_PROCESS : RTL_ENCRYPT_OPTION_SAME_LOGON;
+
+    /* Convert the HEX string to binary ('-cp1:<data>' */
+    StrData = Param + 5;
+    for (i = 0; i < 32; i++)
+    {
+        //Buffer[i] = ConvertHex8(StrData[2 * i], StrData[2 * i + 1]);
+        unsigned x;
```
Comment (SergeGautherie) [inline modules/rostests/apitests/advapi32/RtlEncryptMemory.c:60]:
```suggestion
        unsigned int x;
```

File: modules/rostests/apitests/advapi32/RtlEncryptMemory.c [hunk 7]
```diff
@@ -9,26 +9,117 @@
...
+
+    printf("TestCrossProcessDecrypt(%s)\n", StrBuffer - 5);
+
+    result = system(CmdLine);
+    ok_int(result, 0);
+}
+
+START_TEST(RtlEncryptMemory)
+{
+    CHAR Buffer[8] = { 0 };
```
Comment (SergeGautherie) [inline modules/rostests/apitests/advapi32/RtlEncryptMemory.c:103]:
```suggestion
    CHAR Buffer[RTL_ENCRYPT_MEMORY_SIZE] = { 0 };
```
And use a `#define BUFFER_SIZE (4 * RTL_ENCRYPT_MEMORY_SIZE)` where it applies.

File: modules/rostests/apitests/advapi32/RtlEncryptMemory.c [hunk 8]
```diff
@@ -9,26 +9,117 @@
...
+    ULONG OptionFlags;
+    PSTR StrData;
+    ULONG i;
+    NTSTATUS Status;
+
+    printf("TestCrossProcessDecrypt(%s)\n", Param);
+
+    OptionFlags = Param[3] == '1' ? RTL_ENCRYPT_OPTION_CROSS_PROCESS : RTL_ENCRYPT_OPTION_SAME_LOGON;
+
+    /* Convert the HEX string to binary ('-cp1:<data>' */
```
Comment (SergeGautherie) [inline modules/rostests/apitests/advapi32/RtlEncryptMemory.c:55]:
```suggestion
    /* Convert the HEX string to binary ('-cpN:<data>' */
```

File: modules/rostests/apitests/advapi32/RtlEncryptMemory.c (hunk 6 above)
Comment (HBelusca) [reply modules/rostests/apitests/advapi32/RtlEncryptMemory.c:60]: wouldn't using UCHAR work (for the scanf, due to the usage of the "%02x" formatter) ?

File: modules/rostests/apitests/advapi32/RtlEncryptMemory.c [hunk 9]
```diff
@@ -9,26 +9,117 @@
...
+    {
+        TestCrossProcessDecrypt(Param);
+        return;
+    }
+
+    TestEncrypt(RTL_ENCRYPT_OPTION_SAME_PROCESS);
+    TestEncrypt(RTL_ENCRYPT_OPTION_CROSS_PROCESS);
+    TestEncrypt(RTL_ENCRYPT_OPTION_SAME_LOGON);
+
+    Status = RtlEncryptMemory(Buffer, 8, 4);
```
Comment (HBelusca) [inline modules/rostests/apitests/advapi32/RtlEncryptMemory.c:120]:
```suggestion
    Status = RtlEncryptMemory(Buffer, sizeof(Buffer), 4);
```

File: modules/rostests/apitests/advapi32/RtlEncryptMemory.c [hunk 10]
```diff
@@ -9,26 +9,117 @@
...
+    ULONG i;
+    NTSTATUS Status;
+
+    printf("TestCrossProcessDecrypt(%s)\n", Param);
+
+    OptionFlags = Param[3] == '1' ? RTL_ENCRYPT_OPTION_CROSS_PROCESS : RTL_ENCRYPT_OPTION_SAME_LOGON;
+
+    /* Convert the HEX string to binary ('-cp1:<data>' */
+    StrData = Param + 5;
+    for (i = 0; i < 32; i++)
```
Comment (HBelusca) [inline modules/rostests/apitests/advapi32/RtlEncryptMemory.c:57]:
```suggestion
    for (i = 0; i < _countof(Buffer); i++)
```

File: modules/rostests/apitests/advapi32/RtlEncryptMemory.c [hunk 11]
```diff
@@ -9,26 +9,117 @@
...
+    RtlCopyMemory(Buffer, TestData, sizeof(Buffer));
+    Status = RtlEncryptMemory(Buffer, sizeof(Buffer), OptionFlags);
+    ok_ntstatus(Status, STATUS_SUCCESS);
+
+    /* Start building a command line */
+    sprintf(CmdLine, "advapi32_apitest.exe RtlEncryptMemory -cp%lu:", OptionFlags);
+    StrBuffer = CmdLine + strlen(CmdLine);
+
+    /* Convert encrypted data into a hex string */
+    for (i = 0; i < 32; i++)
```
Comment (HBelusca) [inline modules/rostests/apitests/advapi32/RtlEncryptMemory.c:90]:
```suggestion
    for (i = 0; i < _countof(Buffer); i++)
```

File: modules/rostests/apitests/advapi32/RtlEncryptMemory.c (hunk 8 above)
Comment (HBelusca) [reply modules/rostests/apitests/advapi32/RtlEncryptMemory.c:55]: Missing closing parenthesis in the comment?

File: modules/rostests/apitests/advapi32/RtlEncryptMemory.c [hunk 12]
```diff
@@ -9,26 +9,117 @@
...
 {
-    static const CHAR TestData[32] = "This is some test Message!!!";
-    CHAR Buffer[32];
+    CHAR DECLSPEC_ALIGN(8) Buffer[32];
     NTSTATUS Status;
 
     /* Size must be aligned to 8 bytes */
-    Status = RtlEncryptMemory(Buffer, 7, RTL_ENCRYPT_OPTION_SAME_PROCESS);
+    RtlCopyMemory(Buffer, TestData, sizeof(Buffer));
+    Status = RtlEncryptMemory(Buffer, 7, OptionFlags);
```
Comment (SergeGautherie) [inline modules/rostests/apitests/advapi32/RtlEncryptMemory.c:21]:
```suggestion
    Status = RtlEncryptMemory(Buffer, RTL_ENCRYPT_MEMORY_SIZE + 1, OptionFlags);
```

Comment (HBelusca) [reply modules/rostests/apitests/advapi32/RtlEncryptMemory.c:21]: Come on @SergeGautherie .... he wants to test **7 bytes**, but now you are changing the size to something else....

Comment (SergeGautherie) [reply modules/rostests/apitests/advapi32/RtlEncryptMemory.c:21]: Changed, though `7` or `9` should make no difference.

File: modules/rostests/apitests/advapi32/RtlEncryptMemory.c (hunk 6 above)
Comment (tkreuzer) [reply modules/rostests/apitests/advapi32/RtlEncryptMemory.c:60]: No. %02x has the type (unsigned) int.

File: modules/rostests/apitests/advapi32/RtlEncryptMemory.c [hunk 13]
```diff
@@ -9,26 +9,117 @@
...
 
     RtlCopyMemory(Buffer, TestData, sizeof(Buffer));
-    Status = RtlEncryptMemory(Buffer, sizeof(Buffer), RTL_ENCRYPT_OPTION_SAME_PROCESS);
+    Status = RtlEncryptMemory(Buffer, sizeof(Buffer), OptionFlags);
     ok_ntstatus(Status, STATUS_SUCCESS);
     ok_int(RtlEqualMemory(Buffer, TestData, sizeof(Buffer)), 0);
-    Status = RtlDecryptMemory(Buffer, sizeof(Buffer), RTL_ENCRYPT_OPTION_SAME_PROCESS);
+    Status = RtlDecryptMemory(Buffer, sizeof(Buffer), OptionFlags);
+    ok_ntstatus(Status, STATUS_SUCCESS);
+    ok_int(RtlEqualMemory(Buffer, TestData, sizeof(Buffer)), 1);
```
[the remark this answers is not in the document]
Comment (tkreuzer) [reply modules/rostests/apitests/advapi32/RtlEncryptMemory.c:40]: No

File: drivers/crypto/ksecdd/crypt.c (hunk 1 above)
Comment (learn-more) [inline drivers/crypto/ksecdd/crypt.c:40]: At one point we might want to add a function like this in one of our internal `debug.h` headers

File: modules/rostests/apitests/advapi32/RtlEncryptMemory.c [hunk 14]
```diff
@@ -9,26 +9,112 @@
...
+    StrBuffer = CmdLine + strlen(CmdLine);
+
+    /* Convert encrypted data into a hex string */
+    for (i = 0; i < sizeof(Buffer); i++)
+    {
+        sprintf(&StrBuffer[2 * i], "%02X", Buffer[i]);
+    }
+
+    result = system(CmdLine);
+    ok_int(result, 0);
```
Comment (learn-more) [inline modules/rostests/apitests/advapi32/RtlEncryptMemory.c:93]: An alternative would be something like:
```
                WCHAR buf[MAX_PATH+40];
                STARTUPINFOW si = { sizeof(si) };
                PROCESS_INFORMATION pi;
                BOOL created;

                StringCchPrintfW(buf, _countof(buf), L"\"%ls\" dispmode %u", path, n);
                created = CreateProcessW(NULL, buf, NULL, NULL, FALSE, 0, NULL, NULL, &si, &pi);
                ok(created, "Expected CreateProcess to succeed\n");
                if (created)
                {
                    winetest_wait_child_process(pi.hProcess);
                    CloseHandle(pi.hThread);
                    CloseHandle(pi.hProcess);
                }
```
where `winetest_wait_child_process` does do some exit code validation (for exceptions, and a timeout of 30 seconds on the process wait)

**Review by learn-more — APPROVED** [review]

File: modules/rostests/apitests/advapi32/RtlEncryptMemory.c (hunk 14 above)
Comment (tkreuzer) [reply modules/rostests/apitests/advapi32/RtlEncryptMemory.c:93]: How's that better?

File: modules/rostests/apitests/advapi32/RtlEncryptMemory.c (hunk 8 above)
Comment (SergeGautherie) [reply modules/rostests/apitests/advapi32/RtlEncryptMemory.c]: > Missing closing parenthesis

Right. Updated suggestion.

File: modules/rostests/apitests/advapi32/RtlEncryptMemory.c (hunk 14 above)
Comment (learn-more) [reply modules/rostests/apitests/advapi32/RtlEncryptMemory.c:93]: system call is blocking, the winetest thing times out so it will not hang indefinitely.

Comment (tkreuzer) [reply modules/rostests/apitests/advapi32/RtlEncryptMemory.c:93]: I really don't see a reason why this would hang indefinately.

**Pull request merged**